### ISSUE TITLE
Internal clients should be allowed to bypass auth plugins

### ISSUE THREAD

**Issue by zeferino90** [emqx/emqx#3225]
#### Environment

- OS: Raspbian GNU/Linux 10
- Erlang/OTP: --
- EMQ: emqx 4.0.1

#### Description

I loaded and set up the http_auth plugin (works well) for external use. On internal listener I put anonymous = true. Basically all my settings are on default. 

I expect to connect with clientid, username and password from external ports and to connect anonymously from internal ports but I need to put the credentials on both cases. From the documentation I understand that the anonymous_connection setting prevent from checking the authorization. Am I right?

**Comment by rahulnyk** [comment]
+1 on this issue. I am facing the same problem.

**Comment by turtleDeng** [comment]
Thank you for your participation
First you can use different zones to distinguish

`` `
zone.external.allow_anonymous = false

zone.internal.allow_anonymous = true
`` `
You can return ignore in the body returned by auth_http, it will use the zone configuration
Please refer to details https://docs.emqx.io/broker/latest/en/plugins.html#http-api-return-value-processing

**Comment by zeferino90** [comment]
Hello,

In my configuration I have 
`allow_anonymous = false
zone.internal.allow_anonymous = true
listener.tcp.external = 0.0.0.0:1883
listener.tcp.internal = 127.0.0.1:11883`


I expect for external zone to use the global setup.
`allow_anonymous = false`
And for internal communication I understand that the zone.internal setup override the global setup. Is the weight of these settings like that?  zone setup > global setup

Now I changed to:
`zone.external.allow_anonymous = false
zone.internal.allow_anonymous = true
listener.tcp.external = 0.0.0.0:1883
listener.tcp.internal = 127.0.0.1:11883`

But the same result If I try:
`> mosquitto_sub -h 127.0.0.1 -p 11883 -t 'patata'
Connection Refused: not authorised.`

**Comment by turtleDeng** [comment]
1. After modifying the configuration, you need to restart the service to take effect
2. Did you enable auth_http?

**Comment by rahulnyk** [comment]
The following config worked for me.
HTTP auth enabled and API response as follows:

> ## Authentication succeeded
> HTTP Status Code: 200
> ## Ignore this certification
> HTTP Status Code: 200
> Body: ignore
> ## Authentication failed
> HTTP Status Code: other than 200

EMQX config file

```
# Global 
allow_anonymous = false
# external
zone.external.allow_anonymous = false
listener.tcp.external = 0.0.0.0:1883
# internal
zone.internal.allow_anonymous = true
listener.tcp.internal = 127.0.0.1:11883
```

With this config, I can 
* Connect anonymously at 11883 from localhost, and 
* Connect on 1883 with HTTP auth from outside my VPC
* Not able to connect anonymously from outside my VPC

Which means the HTTP auth plugin does not honor the zone settings. I have to "ignore" the HTTP auth request to reach the zone-specific anonymous settings. 

In my opinion, it would be better if setting up the internal zone to allow the anonymous connection itself bypasses the HTTP auth.

**Comment by zeferino90** [comment]
> 1. After modifying the configuration, you need to restart the service to take effect

yes
> 2. Did you enable auth_http?

yes

Ok I understand the workaround but I agree with @rahulnyk the listener's configuration have to be more important than a the plugin. What if you can't change the behaviour of the http_auth endpoint? (or whatever you use for auth in other plugins).

I understand the zones are a way to create independent configurations inside the same broker, even I miss a way to setup plugin auth per zone, so every zone could use different http endpoints

**Comment by HJianBo** [comment]
> In my opinion, it would be better if setting up the internal zone to allow the anonymous connection itself bypasses the HTTP auth.

Agree that!

I prefer to add an option to allow some a zone of clients to bypass the auth plugins. For example:

```
## Allow the clients for a zone to bypass authentication step
##
## Value: true | false
## Default: false
zone.internal.bypass_auth_plugins = true
```

**Comment by HJianBo** [comment]
Hi guys, we have implemented it and released in the v4.0.3
